ADD: 메인페이지 뷰 추가

gyuseok-dev opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- Gotchapedia 메인페이지용 뷰 작성
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 사용 모델: User, Star, Movie
- 유저가 별점을 매긴 영화중 상위 10개의 영화 정보를 반환
- Postman 테스트
![image](https://user-images.githubusercontent.com/69829621/99653973-0ceca700-2a9d-11eb-980d-5309ebad2cbc.png)
- 프로트엔드와 연동 테스트
![image](https://user-images.githubusercontent.com/69829621/99654266-67860300-2a9d-11eb-8c0a-9fa2497ed9c2.png)

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
models.py migrations

Commit:
[add] users/views.py 작성

gyuseok-dev committed:
ADD: 메인페이지용 뷰 추가
	- 팀원이 각자 좋아하는 영화를 출력
	- 팀원들을 실제 유저로 등록하여 별점을 매기는 MOCK 작업 진행 예정
	- 유저가 별점을 매긴 영화 중 상위 10개의 영화 정보를 리턴

gyuseok-dev committed:
MODIFY: 메인페이지 뷰 수정

- UserFavoriteView 메소드 post에서 get으로 변경
- 쿼리 스트링 인자를 유저이름에서 유저 ID로 변경
- 사용자에게 순위의 범위를 지정할 수 있는 range 추가
- 전달받는 인자가 숫자가 아닐 때 에러핸들링 추가

devops-andrew commented on movie/views.py:
```diff
@@ -1,3 +1,33 @@
+class UserFavoriteView(View):
+    def get(self, request):
+        account_id = request.GET.get('id')
+        list_range = request.GET.get('range')
```
```suggestion
        list_range = request.GET.get('range', LIMIT )
```
코드 상단에 LIMIT라는 상수를 지정하시고, range에 값이 없을시 대입할 기본값을 지정해주세요.

devops-andrew commented on movie/views.py:
```diff
@@ -1,3 +1,33 @@
+
+class UserFavoriteView(View):
+    def get(self, request):
+        account_id = request.GET.get('id')
```
```suggestion
        account_id = request.GET.get('id')
```
아래의 validation을 보니 이뷰는 어카운트 아이디가 들어오지 않으면 아에 데이터 출력이 없는 뷰 인가요?

gyuseok-dev replied:
맞습니다. 테스트해봤을 때 유저가 없으면 빈 JSON 출력입니다. 에러처리 하는 게 좋을까요?

devops-andrew replied:
네 지금 아래 코드에 에러로 출력하고 있어서 여쭤본거 였습니다.  아래 Key 에러처리를 지워주시고, 빈 JSON이 나가도록 해주세요.

devops-andrew commented on movie/views.py:
```diff
@@ -1,3 +1,33 @@
+            movies = [star.movie for star in stars]
+
+            content = {
+                'message': 'SUCCESS',
```
데이터가 출력된다면, 메세지는 제공할 필요가 없습니다.

devops-andrew commented on movie/views.py:
```diff
@@ -1,3 +1,33 @@
+                            'rank': i+1,
+                            'date': f'{movie.opening_at.year} . {movie.country}'
+                        } for i, movie in enumerate(movies)]
+            }
```
```
stars  = Star.objects.filter(user_id=int(account_id)).order_by('-point')[:int(list_range)]
            movies = [star.movie for star in stars]

            content = {
                'message': 'SUCCESS',
                'data': [{
                            'imageURL': movie.main_image,
                            'title': movie.name,
                            'rate': f'{stars[i].point}',
                            'rank': i+1,
                            'date': f'{movie.opening_at.year} . {movie.country}'
                        } for i, movie in enumerate(movies)]
            }
```
여전히 리팩토링의 여지가 있습니다.
랭크는 꼭 만들어야 하나요? 순서가 보장해주지 않나요?
```
rank_data =  [
    {  
        'image_url': star.movie.main_image,
        'title'          : star.movie.name,
        'rate'          : star.point,
        'date'         : f'{star.movie.opening_at.year},{star.movie.country}'
       } for star in  Star.objects.filter(user_id=int(account_id)).order_by('-point')[:int(list_range)]
```
직접 수행한게 아니라서 작동할지 모르겠지만 Let's Try!

gyuseok-dev replied:
질문있습니다.
1. 제안해주신 것처럼 작성하면 star.movie.main_image와 star.movie.name을 사용할 때 데이터 베이스에서 호출이 일었나지 않나요? 적게 호출하는 것은 좋은 것이라고 생각해서 movie의 쿼리를 먼저 호출하고 반복문을 작성하였습니다.

데이터 베이스 호출 횟수가 동일하거나 효율이 크게 다르지 않다면 굳이 enumerate도 사용할 필요도 없고 훨씬 간편한 구문이라고 생각됩니다. 반영하도록 하겠습니다.

gyuseok-dev replied:
랭크의 경우 프로트엔드 담당자와 결정한 것이여서 다시 상의한 후 반영하도록 하겠습니다. 감사합니다!

devops-andrew reviewed (changes requested):
확인 부탁드립니다!

gyuseok-dev committed:
MODIFY: 메인 페이지 뷰 수정

- 쿼리 스트링 사용
- 전달인자
	- id: 원하는 사용자 아이디
	- range: 범위
- 쿼리 셋 더 단순한 구조로 변경

gyuseok-dev committed:
MODIFY: 메인 페이지 뷰 수정

- 디테일 뷰에서 전달 받을 영화 id 값 추가
- rank 요소 제거
- select_related 사용해서 refactoring
- 테스트 완료

gyuseok-dev committed:
MODIFY: 메인페이지 뷰 리스트 범위 수정
- 별점 평가한 개수가 range_list에 못미치면 out_of_range raise되는 문제 수정
- range_list 와 별점 평가한 개수중 최소값 w사용

devops-andrew commented on movie/views.py:
```diff
@@ -1,3 +1,36 @@
+            return JsonResponse({'message': 'KEY_ERROR'}, status=400)
+
+        try:
+            stars = Star.objects.select_related('movie').filter(user_id=int(account_id)).order_by('-point').all()
```
```suggestion
            stars = Star.objects.select_related('movie').filter(user_id=int(account_id)).order_by('-point')[:LIMIT]
```
all()은 제거해주시고 LIMIT는 위쪽에 상수로 선언해서 이용해주시기 바랍니다.

devops-andrew commented on movie/views.py:
```diff
@@ -1,3 +1,36 @@
+                    'title': star.movie.name,
+                    'rate': star.point,
+                    'date': f'{star.movie.opening_at.year} . {star.movie.country}'
+                 } for star in stars[:range]]
```
위의 수정을 하면 range는 필요없습니다.
```suggestion
                 } for star in stars]
```

gyuseok-dev committed:
MODIFY: 영화 리스트 GET 기능 수정

- Class 이름 변경(UserFavoritView -> MovieUserView)
- Query Parameter로 유저 아이디 값 전달 시 유저가 별점을 준 영화 리스트 Respond
- id가 없으면 전체 영화 리스트 출력

devops-andrew commented:
LGTM
